src/kbc_modules/cc_kbc/crypto/mod.rs

```diff
@@ -0,0 +1,130 @@
+// Copyright (c) 2022 Alibaba Cloud
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+
+use aes_gcm::aead::{Aead, NewAead};
+use aes_gcm::Aes256Gcm;
+use anyhow::*;
+use rsa::pkcs8::EncodePublicKey;
+use rsa::{PaddingScheme, RsaPrivateKey, RsaPublicKey};
+use serde::{Deserialize, Serialize};
+use sha2::{Digest, Sha384};
+
+const RSA_ALGORITHM: &str = "rsa-pkcs1v15";
+const RSA_PUBKEY_LENGTH: usize = 2048;
+const NEW_PADDING: fn() -> PaddingScheme = PaddingScheme::new_pkcs1v15_encrypt;
+
+const AES_GCM_256_ALGORITHM: &str = "aes-gcm-256";
+
+// The key inside TEE to decrypt confidential data.
+#[derive(Debug, Clone)]
+pub struct TeeKey {
+    private_key: RsaPrivateKey,
+    public_key: RsaPublicKey,
+}
+
+// The public key of TEE.
+#[derive(Serialize, Deserialize, Debug, Clone)]
+pub struct TeePubKey {
+    algorithm: String,
+    #[serde(rename = "pubkey-length")]
+    pubkey_length: usize,
+
+    // public key string in PEM format.
+    pubkey: String,
+}
+
+impl TeeKey {
+    pub fn new() -> Result<TeeKey> {
+        let mut rng = rand::thread_rng();
+
+        let private_key = RsaPrivateKey::new(&mut rng, RSA_PUBKEY_LENGTH)?;
+        let public_key = RsaPublicKey::from(&private_key);
+
+        Ok(TeeKey {
+            private_key,
+            public_key,
+        })
+    }
+
+    // Export TEE public key as specific structure.
+    pub fn export_pubkey(&self) -> Result<TeePubKey> {
+        let pem_line_ending = rsa::pkcs1::LineEnding::default();
+        let pubkey_pem_string = self.public_key.to_public_key_pem(pem_line_ending)?;
+
+        Ok(TeePubKey {
+            algorithm: RSA_ALGORITHM.to_string(),
+            pubkey_length: RSA_PUBKEY_LENGTH,
+            pubkey: pubkey_pem_string,
+        })
+    }
+
+    // Use TEE private key to decrypt cipher text.
+    pub fn decrypt(&self, cipher_text: Vec<u8>) -> Result<Vec<u8>> {
+        let padding = NEW_PADDING();
```

Comment:
I'm a little confused about this const fn. Why is it needed?

Reply:
The idea was to keep the definitions of `RSA_ALGORITHM` and `NEW_PADDING` next to each other, because they are coupled (should be changed together, if we change them in the future). They are at the top of this file:

```rust
const RSA_ALGORITHM: &str = "rsa-pkcs1v15";
const RSA_PUBKEY_LENGTH: usize = 2048;
const NEW_PADDING: fn() -> PaddingScheme = PaddingScheme::new_pkcs1v15_encrypt;
```

I agree that the solution (which I suggested) is not very elegant or "rusty" ...

Suggestions welcome! Maybe the original was better...

Reply:
@fitzthum This is a necessary work in the standard usage of `rsa` crate. For details, you can refer to [this](https://docs.rs/rsa/0.6.1/rsa/index.html).

Reply:
Ok, makes sense. I've only seen people use const functions when they want to use a function to define another const, but this seems reasonable.

Reply:
I'm not sure whether other algorithms added in the future will need padding (at least RSA needs it at present), so maybe we can consider code optimization here when new algorithms are added?

Reply:
@jialez0 - Yep - it was just an idea. And we could of course make the handler optional:

```rust
handler: Option<fn() -> PaddingScheme>,
```